### ISSUE TITLE
[QST] Binary Classification - Dataloaders

### ISSUE THREAD

**Issue by jacobdineen** [NVIDIA-Merlin/Transformers4Rec#405]
# ❓ Questions & Help

## Details

I wanted to start off by saying thank you for a great library! 

The NVTabular dataloader yields a dictionary of a batch of samples with {feature_name: tensors) as input into the model - These inputs are drawn from the schema file and their respective tags. For instance, a continuous tagged feature could flow through the input module and be projected into a latent vector, and the cat features flow through the categorical module before they are aggregated (see code snippet below).

```
import torch
from transformers4rec import torch as tr
max_sequence_length, d_model = 50, 80


input_module = tr.TabularSequenceFeatures.from_schema(
    schema,
    max_sequence_length=max_sequence_length,
    categorical_tags=["categorical"],
    continuous_tags= ["continuous"],
    labels=["target"],
    continuous_projection=[d_model, d_model],
    aggregation="concat",
    d_output=d_model,
)



prediction_task = tr.BinaryClassificationTask(target_name="target" 
                                                          ,summary_type="last")

transformer_config = tr.XLNetConfig.build(
    d_model=d_model, n_head=8, n_layer=8, dropout = 0.3, total_seq_length=max_sequence_length, summary_activation="tanh", 
)

body = tr.SequentialBlock(
    input_module, 
    tr.MLPBlock([d_model,d_model,d_model], dropout = 0.3),
    tr.TransformerBlock(transformer_config),
    )

head = tr.Head(
    body,
    prediction_task,
    inputs=input_module,
)

model = tr.Model(head)
```

An example of a batch of data using the default NVTabularDataLoader yields only the input features to the model, rather than the expected X,y batch: 
```
next(iter(train.get_eval_data_loader()))

{'variant-list_seq': tensor([[ 7,  7,  0,  ...,  0,  0,  0],
        [11, 11,  0,  ...,  0,  0,  0],
        [ 2,  2,  0,  ...,  0,  0,  0],
        ...,
        [ 3,  3,  0,  ...,  0,  0,  0],
        [ 3,  3,  0,  ...,  0,  0,  0],
        [ 8,  8,  0,  ...,  0,  0,  0]], device='cuda:0'), 'et_dayofweek_sin-list_seq': tensor([[0.9749, 0.9749, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
        [0.9749, 0.9749, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
        [0.9749, 0.9749, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
        ...,
        [0.9749, 0.9749, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
        [0.9749, 0.9749, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
        [0.9749, 0.9749, 0.0000,  ..., 0.0000, 0.0000, 0.0000]],
       device='cuda:0')}

```

 Assuming that the schema file tags a feature as "target", the prediction head is specified to use this feature to compute the BCELoss (by default):

`prediction_task = tr.BinaryClassificationTask(target_name="target"  ,summary_type="last")`

Where is the 'target` feature flowing from in this example, or any BCtask? This could be a potential reason why calling trainer.evaluate() does not currently work, as referenced in 2 separate issues, but more than that, I am confused on how/if the target feature is currently utilized during training for the BCTask if the dataloader is not yielding it. For NextItemPrediction, there is no target, and the data is masked, but for the BCtask, the target should be directly used to calculate the loss / backwards pass.

**Comment by sararb** [comment]
Thank you @jacobdineen for your interest in the Transformers4Rec library. The `Trainer` class and its utilities were implemented for session-based and sequential tasks (i.e NextItemPredictionTask). In this case, the input to the model is a dictionary of tensors and the targets are computed on the fly (via masking class). So, the class does not support tasks with explicit targets like BinaryClassification. 

For that, we have a naive PyTorch implementation of the `fit` and `evaluate` methods that you can use with your model. You will need to define the nvtabular data-loader as follow: 
```
from transformers4rec.torch.utils.data_utils import NVTabularDataLoader
nvt_loader_train = NVTabularDataLoader.from_schema(
    schema=schema,
    paths_or_dataset=$PATHS_TO_TRAIN_DATA, 
    batch_size=32, 
    collate_fn=lambda x: x,
    max_sequence_length=20)
```
Then you call fit and evaluate as follow: 
```
model.fit(nvt_loader_train, num_epochs=5)
model.evaluate(nvt_loader_train)
```


Also, we have recently released the [Merlin Models](https://github.com/NVIDIA-Merlin/models) library that supports the implementation of most common recsys architectures (DLRM, DCN, Two-Tower) but also offers all the necessary building blocks to train with different inputs, architectures, and prediction tasks. You can check the example notebooks examples [here](https://github.com/NVIDIA-Merlin/models/tree/main/examples). We are also planning to port the sequential and session-based models from Transformers4Rec (GRU4Rec, Transformers…) in future releases. 
The current implementation supports Tensorflow API and the PyTorch API is in the roadmap.

**Comment by jacobdineen** [comment]
Thanks @sararb! I will check out the test case and try to build it up like you've noted. Out of curiosity - Even when the task head is specified as BC rather than NextItemPred, the trainer.train() call compiles - even when the model has no notion of the target from the NVTDataloaders. Is this still just doing masking under the hood?


Tagging the binary response as 'target' in the schema and then specifiying targets in the dataloader yields the batch (X,y), for others looking into this:

```
from transformers4rec.torch.utils.data_utils import NVTabularDataLoader
nvt_loader_train = NVTabularDataLoader.from_schema(
    schema=schema,
    paths_or_dataset=train_paths, 
    targets = ['target'],
    batch_size=32, 
    collate_fn=lambda x: (x),
    max_sequence_length=50)
```

**Comment by jacobdineen** [comment]
This can be closed. Thanks @sararb for the help here! This style of pipeline works & may be useful in a future tutorial. Thanks again to the T4rec/Merlin team:


```python
def get_dataloader(schema, path:str, batch_size:int=256, target:list=['target']):
    nvt_loader = NVTabularDataLoader.from_schema(
        schema=schema,
        paths_or_dataset=path, 
        targets = target,
        batch_size=batch_size, 
        collate_fn=lambda x: x,
        max_sequence_length=20)
    return nvt_loader

# Define input module to process tabular input-features and to prepare masked inputs
input_module = tr.TabularSequenceFeatures.from_schema(
    schema,
    max_sequence_length=20,
    continuous_projection=[64,64,64],
    aggregation="concat",
    d_output=64,
)


metrics = (
            tm.Precision(),
            tm.Recall(),
            tm.Accuracy(),
            tm.F1(),
            tm.ROC(),
            )

prediction_task = tr.BinaryClassificationTask("target", summary_type='last', loss = torch.nn.BCEWithLogitsLoss(), metrics=metrics)


transformer_config = tr.XLNetConfig.build(
    d_model=64, n_head=8, n_layer=8, dropout = 0.3, total_seq_length=20
)

#Get the end-to-end model 
model = transformer_config.to_torch_model(input_module, prediction_task)
model = model.cuda()

start_time_window_index =1
final_time_window_index = 2
for time_index in range(start_time_window_index, final_time_window_index):
    time_index_train = time_index
    time_index_eval = time_index + 1
    
    train_path = glob.glob(os.path.join('preproc_sessions_by_day', f"{time_index_train}/train.parquet"))
    eval_path = glob.glob(os.path.join('preproc_sessions_by_day', f"{time_index_eval}/valid.parquet"))
    
    train_loader = get_dataloader(schema, train_path, batch_size=256, target=['target'])
    eval_loader = get_dataloader(schema, eval_path, batch_size=256, target=['target'])
    
    print('training')
    model.fit(train_loader, num_epochs=1, amp = True, verbose = False)
    print('evaluating')
    metrics = model.evaluate(train_loader)
    wipe_memory()
```
